### DISCUSSION TITLE
Interfacing with the 1520 printer plotter using DurexForth

### DISCUSSION THREAD

**Discussion by gdm413229** [jkotlinski/durexforth#384] [General]
I'm currently writing a 1520 interface library for DurexForth, and am currently struggling with the provided low-level device I/O routines. Any guidance is welcome and greatly appreciated.

**Comment by Whammo** [comment]
There are lots of examples for file handling I've left in discussions. Here is another.
  You'll want to look at:
  https://github.com/jkotlinski/durexforth/blob/master/forth_src/io.fs
  and
 https://github.com/jkotlinski/durexforth/blob/master/forth_src/open.fs
  
  Device number is presumed to be DEVICE, so you only need file number and secondary address for OPEN.
 
 6 DEVICE sets device number for plotter.


```forth
require io
: less \ less filename
parse-name ?dup if
else drop s" less"
then
8 8 open ioabort
clrchn
8 chkin ioabort
chrin chrin 2drop page  \ load address, clr screen
begin more chrin emit
readst until
clrchn 8 close
;

**Reply by Whammo** [reply]
Interesting- there doesn't seem to be a word for CHROUT in the io library.
  This will be helpful and get you where you want to go.
  
 
 ```forth
: chrout ( b -- )
ar c! $ffd2 sys ;
```

**Reply by gdm413229** [reply]
Is the `ar` thing referencing the 6502/6510/8502's accumulator? That word might come in handy for doing low-level plotter functions through the KERNAL.

**Reply by Whammo** [reply]
We've hijacked the basic sys command and made available AR, XR, YR, and SR to preload for SYS, and they will be valid upon rts.

**Comment by Whammo** [comment]
Also, EMIT works the same as CHROUT even when output is redirected.
  Similarly TYPE can be used for strings
  TYPE  ( caddr u -- )

**Comment by Whammo** [comment]
This is my best untested guess.

BASIC
```basic
open 1,6,0
print#1, "hello!"
CLOSE 1
```
FORTH
```forth
include io
: hello
6 device
0 0 1 0 open ioabort \ no filename, channel 1, sa 0
clrchn
s" hello!"
1 chkout ioabort type cr
clrchn
1 close ;
```

Edit: oops ioabort
